element_electrode_localization/electrode_localization.py

```diff
@@ -1,48 +1,211 @@
+
+        if len(channel_locations_files) == 1:
+            corresponding_shanks = [0]
+            if len(shanks) != 1:
+                raise ValueError(
+                    'Only 1 file found ({}) for a {}-shank probe'.format(
+                        channel_locations_files[0].name, len(shanks)))
+            if 'shank' in channel_locations_files[0].stem \
+               and channel_locations_files[0].stem[-1] != 1:
+                raise ValueError('The electrode-location file found ({}) is '
+                                 + 'unexpected for this 1-shank probe')
+        else:
+            if len(channel_locations_files) != len(shanks):  # ensure 1 file per shank
+                raise ValueError(f'{len(channel_locations_files)} files found for a '
+                                 + f'{len(shanks)}-shank probe')
+            corresponding_shanks = [int(f.stem[-1]) for f in channel_locations_files]
+
+        # Insertion
+        self.insert1(key)
+
+        for channel_locations_file, shank_no in zip(
+                channel_locations_files,
+                corresponding_shanks):
+
+            log.debug(f'loading channel locations from {channel_locations_file}')
+            with open(channel_locations_file, 'r') as fh:
+                chn_loc_raw = json.loads(fh.read())
+
+            chn_loc_data = {'origin': chn_loc_raw['origin']}
+
+            if len(chn_loc_data['origin'].keys()) > 1:
+                log.error('More than one origin region found ({}). skipping.'.format(
+                    chn_loc_data['origin']))
+                raise ValueError('More than one origin region found '
+                                 + f'({chn_loc_data["origin"]})')
+
+            # ensuring channel data is sorted;
+            chn_loc_keymap = {int(k.split('_')[1]): k for k
+                              in chn_loc_raw.keys() if 'channel_' in k}
+
+            chn_loc_data['channels'] = np.array(
+                [tuple(chn_loc_raw[chn_loc_keymap[k]].values()) for k in sorted(
+                    chn_loc_keymap.keys())],
+                dtype=[
+                    ('x', float), ('y', float), ('z', float),
+                    ('axial', float), ('lateral', float),
+                    ('brain_region_id', int), ('brain_region', object)])
+
+            # get/scale xyz positions
+            pos_xyz_raw = np.array([chn_loc_data['channels'][i]
+                                    for i in ('x', 'y', 'z')]).T
+
+            pos_origin = chn_loc_data['origin'][
+                list(chn_loc_data['origin'].keys())[0]]
+
+            pos_xyz = np.copy(pos_xyz_raw)
+
+            # by adjusting xyz axes & offsetting from origin position
+            # in "pos_xyz_raw", x-axis and y-axis are swapped, correcting for that below
+            pos_xyz[:, 0] = pos_origin[1] - pos_xyz_raw[:, 1]
+            pos_xyz[:, 1] = pos_origin[0] + pos_xyz_raw[:, 0]
+            pos_xyz[:, 2] = pos_origin[2] - pos_xyz_raw[:, 2]
```

Comment:
Hi @ttngu207, why is the shifting by the origin done differently for the x/z axes versus y axis (i.e. why subtraction versus addition)?  Thanks.

Reply:
From my understanding, the electrode coordinates in `channel_locations.json` are in Bregma coordinate space.  These locations then have to be transformed to CCF coordinate space.  See [iblapps wiki](https://github.com/int-brain-lab/iblapps/wiki/4.-Preparing-data-for-ephys-GUI#coordinate-systems).  Also see [IBL Ephys Atlas GUI](https://github.com/int-brain-lab/iblapps/tree/master/atlaselectrophysiology).

Reply:
I am likely overlooking something, but the `map-ephys` ingestion script is adjusting the axes slightly differently.

[map-ephys.pipeline.ingest.histotology lines 268-271](https://github.com/datajoint-company/map-ephys/blob/6db75a6b65c6850d27124a7d08c3e44c0b303531/pipeline/ingest/histology.py#L268-L271)

```python
# by adjusting xyz axes & offsetting from origin position
pos_xyz[:, 0] = pos_origin[0] + pos_xyz_raw[:, 0]
pos_xyz[:, 1] = pos_origin[2] - pos_xyz_raw[:, 2]
pos_xyz[:, 2] = pos_origin[1] - pos_xyz_raw[:, 1]
```

Reply:
Updated my comments above.

Reply:
MAP ephys defines the x,y,z axes differently, for historical reason. https://github.com/mesoscale-activity-map/map-ephys/blob/bed146d45739f5b134da01a7b7f3cd0f930863a5/pipeline/ccf.py#L41
```
@schema
class CCF(dj.Lookup):
    # Note: normative CCF has {x: AP, y: DV, z: ML}
    definition = """
    # Common Coordinate Framework
    -> CCFLabel
    ccf_x   :  int   # (um)  Left-to-Right (ML axis)
    ccf_y   :  int   # (um)  Superior-to-Inferior (DV axis)
    ccf_z   :  int   # (um)  Anterior-to-Posterior (AP axis)
    index(ccf_y, ccf_z)
    """
```

Whereas, in this element, our xyz definition aligns with Allen's atlas.

Reply:
Thanks for clarifying @ttngu207.

After discussion with @ttngu207, for the data tested this function appears to be correct as the electrodes show up in the correct space.  Will need to test for other cases and will continue this work in #9.